packages/hydrogen/src/csp/csp.ts

```diff
@@ -0,0 +1,88 @@
+import {
+  type ComponentType,
+  createContext,
+  createElement,
+  type ReactNode,
+  useContext,
+} from 'react';
+import cspBuilder from 'content-security-policy-builder';
+import {generateNonce} from './nonce';
+
+export const NonceContext = createContext<string | undefined>(undefined);
+export const NonceProvider = NonceContext.Provider;
+
+export const useNonce = () => useContext(NonceContext);
+
+type ContentSecurityPolicy = {
+  /** A randomly generated nonce string that should be passed to any custom `script` element */
+  nonce: string;
+  /** The content security policy header */
+  header: string;
+  NonceProvider: ComponentType<{children: ReactNode}>;
+};
+
+/**
+ * @param directives - Pass custom [content security policy directives](https://content-security-policy.com/). This is important if you load content in your app from third-party domains.
+ */
+export function createContentSecurityPolicy(
+  directives: Record<string, string[] | string | boolean> = {},
+): ContentSecurityPolicy {
+  const nonce = generateNonce();
+  const header = createCSPHeader(nonce, directives);
+
+  const Provider = ({children}: {children: ReactNode}) => {
+    return createElement(NonceProvider, {value: nonce}, children);
+  };
+
+  return {
+    nonce,
+    header,
+    NonceProvider: Provider,
+  };
+}
+
+function createCSPHeader(
+  nonce: string,
+  directives: Record<string, string[] | string | boolean> = {},
+): string {
+  const nonceString = `'nonce-${nonce}'`;
+  const defaultDirectives: Record<string, string[] | string | boolean> = {
+    baseUri: ["'self'"],
+    defaultSrc: [
+      "'self'",
+      nonceString,
+      'https://cdn.shopify.com',
+      // Used for the Customer Account API
+      'https://shopify.com',
+    ],
+    frameAncestors: ['none'],
+    styleSrc: ["'self'", "'unsafe-inline'", 'https://cdn.shopify.com'],
+    connectSrc: ['self', 'https://monorail-edge.shopifysvc.com'],
```

Comment:
Yeah I went back and forth on that. I decided to keep make sure the nonce is always merged in, because that's the main value prop we are providing. So if someone doesn't want the nonce, they should just do their own CSP manually. But if we automatically merge domains as well? If they are deploying off oxygen, there's a high chance that they wouldn't want the shopify domains at all.